ggml.c

```diff
@@ -1417,6 +1417,93 @@ inline static void ggml_vec_dot_q4_0(const int n, float * restrict s, const void
 #else
 #error "not implemented for QK"
 #endif
+#elif defined(__AVX512F__)
+inline __m256i bytesFromNibbles( const uint8_t* rsi ){
```

Comment:
This should only be defined once (`#if defined(__AVX2__) || defined(__AVX512F__)`)

Reply:
Should we be using nested functions?

Reply:
Good call on re-using the `byteFromNibbles`.  Code has been updated.

As for nested functions, I'm not sure what you mean.  Do you mean the use of `process_one_block`?  I originally considered implementing it as a macro, but that made the code very messy with all the backslashes.

I also tried using `__attribute__((always_inline))` to make extra sure it always inlines, but that produced warnings.  I'm not super familiar with C/C++, so if there is a better way to achieve that I'm happy to use it.

That being said, this is always inlining for me and it would be very surprising if it were to fail to inline for pretty much any compiler, as long as optimizations are enabled.

Reply:
`process_one_block` is nested inside `ggml_vec_dot_q4_0`, that's not really standard C.

`ggml_vec_dot_q4_0` is at a point now where it should probably be refactored into a specific function for each kind of optimization.

Reply:
Ah, I see.  I wasn't aware that wasn't standard in C.

I've split it out and renamed it to make it apparent that it's used as a helper function for the `dot_q4_0` function.